pkg/executor/adapter_slow_log.go

```diff
@@ -0,0 +1,183 @@
+		},
+	}
+}
+
+// PrepareSlowLogItemsForRules builds a SlowQueryLogItems containing only the fields referenced by current session's SlowLogRules.
+// These pre-collected fields are later used for matching SQL execution details against the rules.
+func PrepareSlowLogItemsForRules(ctx context.Context, seVars *variable.SessionVars) *variable.SlowQueryLogItems {
+	items := &variable.SlowQueryLogItems{}
+	for field := range seVars.SlowLogRules.AllConditionFields {
+		gs := variable.SlowLogRuleFieldAccessors[field]
+		if gs.Setter != nil {
+			gs.Setter(ctx, seVars, items)
+		}
+	}
+
+	return items
+}
+
+// Match checks whether the given SlowQueryLogItems satisfies any of the
+// current session's slow log trigger rules.
+//
+// Matching is evaluated in two levels of logical relationship:
+//   - Rules are combined with OR: if any rule is satisfied, the function returns true.
+//   - Conditions inside a rule are combined with AND: all conditions must be met for that rule.
+//
+// Returns true if any rule matches, false otherwise.
+func Match(seVars *variable.SessionVars, items *variable.SlowQueryLogItems) bool {
+	rules := seVars.SlowLogRules
+	// Or logical relationship
+	for _, rule := range rules.Rules {
+		matched := true
+
+		// And logical relationship
+		for _, condition := range rule.Conditions {
+			accessor := variable.SlowLogRuleFieldAccessors[strings.ToLower(condition.Field)]
+			if ok := accessor.Match(seVars, items, condition.Threshold); !ok {
+				matched = false
+				break
+			}
+		}
+		if matched {
+			return matched
+		}
+	}
+
+	return false
+}
+
+// CompleteSlowLogItemsForRules fills in the remaining SlowQueryLogItems fields
+// that are relevant to triggering the current session's SlowLogRules.
+// It's exporting for testing.
+func CompleteSlowLogItemsForRules(ctx context.Context, seVars *variable.SessionVars, items *variable.SlowQueryLogItems) {
+	if items == nil {
+		return
+	}
+
+	for field, sg := range variable.SlowLogRuleFieldAccessors {
+		if _, ok := seVars.SlowLogRules.AllConditionFields[field]; ok {
+			continue
+		}
+
+		if sg.Setter != nil {
+			sg.Setter(ctx, seVars, items)
+		}
+	}
+}
+
+// SetSlowLogItems fills the remaining fields of SlowQueryLogItems after SQL execution.
+func SetSlowLogItems(a *ExecStmt, txnTS uint64, hasMoreResults bool, items *variable.SlowQueryLogItems) {
+	CompleteSlowLogItemsForRules(a.GoCtx, a.Ctx.GetSessionVars(), items)
+
+	sessVars := a.Ctx.GetSessionVars()
+	stmtCtx := sessVars.StmtCtx
+	var indexNames string
+	if len(stmtCtx.IndexNames) > 0 {
+		// remove duplicate index.
+		idxMap := make(map[string]struct{})
+		buf := bytes.NewBuffer(make([]byte, 0, 4))
+		buf.WriteByte('[')
+		for _, idx := range stmtCtx.IndexNames {
+			_, ok := idxMap[idx]
+			if ok {
+				continue
+			}
+			idxMap[idx] = struct{}{}
+			if buf.Len() > 1 {
+				buf.WriteByte(',')
+			}
+			buf.WriteString(idx)
+		}
+		buf.WriteByte(']')
+		indexNames = buf.String()
+	}
```

Comment:
I see this code block is duplicated in adapter.go. Can we extract it into a separate function?

Reply:
Yes, this function has not been called yet. 
In the next PR, the code in adapter.go will be directly replaced.